static-analysis/teammates-macker.xml

```diff
@@ -1,45 +1,336 @@
+                <deny>
+                    <from pattern="productionCode" />
+                    <to class="teammates.storage.entity.**" />
+                </deny>
                 <allow>
-                    <!-- allow dependency from inner class -->
-                    <from class="${testCasesPackage}.**$**" />
+                    <from>
+                        <include class="teammates.storage.entity.**" />
+                        <include class="teammates.storage.api.**" />
+                        <include class="teammates.common.datatransfer.attributes.**" />
+                    </from>
                 </allow>
                 <allow>
-                    <!-- allow dependency to inner class -->
-                    <to class="${testCasesPackage}.**$**" />
+                    <!-- Violation that needs fixing -->
+                    <from class="teammates.logic.core.FeedbackResponsesLogic" />
+                    <to class="teammates.storage.entity.FeedbackResponse" />
                 </allow>
-                <allow>
-                    <!-- allow dependency to base class -->
-                    <to class="${testCasesPackage}.**Base**" />
-                </allow>
-                <allow>
-                    <!-- allow dependency from Feedback*QuestionUiTest to FeedbackQuestionUiTest -->
-                    <from class="${testCasesPackage}.browsertests.Feedback*QuestionUiTest" />
-                    <to class="${testCasesPackage}.browsertests.FeedbackQuestionUiTest" />
-                </allow>
-            </deny>
-        </access-rule>
+            </access-rule>
+        </ruleset>
+        <ruleset name="StorageComponent: storage::search should only be accessed by storage::api">
+            <access-rule>
+                <deny>
+                    <from pattern="productionCode" />
+                    <to class="teammates.storage.search.**" />
+                    <allow>
+                        <from>
+                            <include class="teammates.storage.api.**" />
+                            <include class="teammates.storage.search.**" />
+                        </from>
+                    </allow>    
+                </deny>
+            </access-rule>
+        </ruleset>
     </ruleset>
-    <ruleset name="Only UI tests can access page object classes">
-        <access-rule>
-            <deny>
-                <from class="teammates.test.**" />
-                <to class="teammates.test.pageobjects.*" />
-                <allow>
-                    <!-- allow dependency within the package -->
-                    <from class="teammates.test.pageobjects.*" />
-                </allow>
-                <allow>
-                    <!-- allow dependency from UI tests -->
-                    <from class="teammates.test.cases.browsertests.*" />
-                </allow>
-            </deny>
-        </access-rule>
+    
+    <ruleset name="Architectural rules for client code">
```

Comment:
I think you removed the `client -> ui` restriction, which is valid. Add it again?